python/nmodl/ode.py

```diff
@@ -344,6 +344,55 @@ def solve_lin_system(
     return code, new_local_vars
 
 
+def finite_difference_step_variable(sym):
+    return f"{sym}_delta_"
+
+
+def discretize_derivative(expr):
+    if isinstance(expr, sp.Derivative):
+        x = expr.args[1][0]
+        dx = sp.symbols(finite_difference_step_variable(x))
+        return expr.as_finite_difference(dx)
+    else:
+        return expr
+
+
+def transform_expression(expr, transform):
+    if expr.args is tuple():
+        return expr
+
+    args = list(transform_expression(transform(arg), transform) for arg in expr.args)
+    return expr.func(*args)
+
+
+def transform_matrix_elements(mat, transform):
+    return sp.Matrix(
+        [
+            [transform_expression(mat[i, j], transform) for j in range(mat.rows)]
+            for i in range(mat.cols)
+        ]
+    )
+
+
+def finite_difference_variables(mat):
+    vars = []
+
+    def recurse(expr):
+        for arg in expr.args:
+            if isinstance(arg, sp.Derivative):
+                var = arg.args[1][0]
+                vars.append((var, finite_difference_step_variable(var)))
+
+    for expr in mat:
+        recurse(expr)
+
+    return vars
```

Comment:
It's admittedly a bit difficult to parse this; how about this alternative (provided I didn't make any typos):

```suggestion
def finite_difference_variables(mat):
    return [
        ((var := arg.args[1][0]), finite_difference_step_variable(var))
        for expr in mat
        for arg in expr.args
        if isinstance(arg, sp.Derivative)
    ]
```

Reply:
ah, we don't need this at all.

Reply:
... and both a buggy, neither actually descends recurses down the AST.